File: src/main/java/org/taktik/icure/logic/CalendarItemLogic.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic;

import org.taktik.icure.db.PaginatedList;
import org.taktik.icure.db.PaginationOffset;
import org.taktik.icure.entities.CalendarItem;
import org.taktik.icure.exceptions.DeletionException;

import java.time.Instant;
import java.util.List;

public interface CalendarItemLogic extends EntityPersister<CalendarItem, String> {

	CalendarItem createCalendarItem(CalendarItem CalendarItem);

	CalendarItem getCalendarItem(String CalendarItemId);

	List<CalendarItem> getCalendarItemByPeriodAndHcPartyId(Long startDate, Long endDate, String HcPartyId);

    List<CalendarItem> getCalendarItemByPeriodAndAgendaId(Long startDate, Long endDate, String agendaId);

    CalendarItem modifyCalendarItem(CalendarItem CalendarItem);

    List<String> deleteCalendarItems(List<String> ids) throws DeletionException;
}

File: src/main/kotlin/org/taktik/icure/asynclogic/CalendarItemLogic.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.asynclogic

import kotlinx.coroutines.flow.Flow
import org.taktik.couchdb.DocIdentifier
import org.taktik.icure.asyncdao.CalendarItemDAO
import org.taktik.icure.entities.CalendarItem

interface CalendarItemLogic : EntityPersister<CalendarItem, String> {
    suspend fun createCalendarItem(calendarItem: CalendarItem): CalendarItem?
    fun deleteCalendarItems(ids: List<String>): Flow<DocIdentifier>
    suspend fun getCalendarItem(calendarItemId: String): CalendarItem?
    fun getCalendarItemByPeriodAndHcPartyId(startDate: Long, endDate: Long, hcPartyId: String): Flow<CalendarItem>
    fun getCalendarItemByPeriodAndAgendaId(startDate: Long, endDate: Long, agendaId: String): Flow<CalendarItem>
    fun listCalendarItemsByHCPartyAndSecretPatientKeys(hcPartyId: String, secretPatientKeys: List<String>): Flow<CalendarItem>

    suspend fun modifyCalendarItem(calendarItem: CalendarItem): CalendarItem?
    fun getGenericDAO(): CalendarItemDAO
    fun getCalendarItems(ids: List<String>): Flow<CalendarItem>
}
